codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +520,49 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
                 try {
-                    connectionState().incrementStreamWindow(-writtenBytes);
-                    incrementStreamWindow(-writtenBytes);
+                    boolean writeError = true;
+                    try {
+                        writing = true;
```

Comment:
> Ad 1) In what scenario would it make sense to do that?

As I said I don't think it is an issue now, but more just pointing out the limitation this introduces. I'm not sure what extension flow controlled frames will do.

> cancel still errors the whole queue.

Looks like we [return](https://github.com/netty/netty/pull/3456/files#diff-d8ca391b616f0c1db6e8f345ca3284feR448) before emptying the queue if `writing` is true, and I don't see where we then later empty the queue after we are done writing.  Perhaps add a unit test?  

> It's in there to prevent a malicious implementation of FlowControlled.write to modify the queue.

Is this just looking forward to extension frames?  Don't we have as part of the interface that the `state must not change`?  Could we just attribute this to a coding error?